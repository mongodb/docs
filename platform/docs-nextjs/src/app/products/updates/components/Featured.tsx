'use client';
import { css, cx } from '@leafygreen-ui/emotion';
import { H3 } from '@leafygreen-ui/typography';
import { Body } from '@leafygreen-ui/typography';
import Card from '@leafygreen-ui/card';
import { palette } from '@leafygreen-ui/palette';
import { Chip } from '@leafygreen-ui/chip';
import type { ProductUpdateEntry } from '../services/contentstack';
import { useRouter } from 'next/navigation';
import { theme } from '@/styles/theme';

const featuredContainerStyle = css`
  max-width: 1440px;
  margin: 48px auto 0;
  padding: 0 24px;
  margin-bottom: 60px;

  @media ${theme.screenSize.mediumAndUp} {
    padding: 0 32px;
  }

  @media ${theme.screenSize.largeAndUp} {
    padding: 0 32px;
  }

  @media ${theme.screenSize['2XLargeAndUp']} {
    padding: 0 8px;
  }
`;

const featuredContentStyle = css`
  display: flex;
  flex-direction: row;
  gap: 16px;
  flex-wrap: wrap;
  justify-content: flex-start;
  margin-top: 26px;

  @media ${theme.screenSize.mediumAndUp} {
    gap: 16px;
  }

  @media ${theme.screenSize.largeAndUp} {
    gap: 24px;
  }
`;

const featuredCardStyle = css`
  flex: 0 0 100%;
  min-width: 0;
  padding: 24px;
  min-height: 360px;

  @media ${theme.screenSize.mediumAndUp} {
    flex: 0 0 calc(50% - 8px);
    padding: 32px;
  }

  @media ${theme.screenSize.largeAndUp} {
    flex: 0 0 calc(33.333% - 16px);
    padding: 40px;
  }
`;

const dateStyle = css`
  font-size: 13px;
  line-height: 20px;
  color: ${palette.gray.dark2};
  margin-bottom: 16px;
`;

const titleStyle = css`
  font-size: 24px;
  line-height: 32px;
  margin-bottom: 8px;
  color: ${palette.black};
`;

const bodyStyle = css`
  font-size: 16px;
  line-height: 24px;
  color: ${palette.black};
  margin-bottom: 44px;
  display: -webkit-box;
  -webkit-line-clamp: 3;
  -webkit-box-orient: vertical;
  overflow: hidden;
`;

const chipsStyle = css`
  display: flex;
  flex-direction: row;
  gap: 8px;
  flex-wrap: wrap;
`;

const chipStyle = css`
  font-size: 16px;
  padding: 2px;
`;

const Featured = ({ updates }: { updates: ProductUpdateEntry[] }) => {
  const router = useRouter();

  // Filter and limit to 3 most recent featured updates
  const limitedFeaturedUpdates = updates
    .sort((a, b) => new Date(a.created_at).getTime() - new Date(b.created_at).getTime())
    .slice(0, 3);

  return (
    <div className={cx(featuredContainerStyle)}>
      <H3>Featured Updates</H3>
      <div className={cx(featuredContentStyle)}>
        {limitedFeaturedUpdates.map((update: ProductUpdateEntry) => {
          const createdAt = update.beamer_created_at || update.created_at;
          const date = new Date(createdAt).toLocaleDateString('en-US', {
            year: 'numeric',
            month: 'long',
            day: 'numeric',
          });

          const slug = update.title
            .toLowerCase()
            .replace(/[^a-z0-9]+/g, '-')
            .replace(/(^-|-$)/g, '');

          return (
            <Card
              key={update.uid}
              className={cx(featuredCardStyle)}
              onClick={() => router.push(`/products/updates/${slug}`)}
            >
              <Body className={cx(dateStyle)}>{date}</Body>
              <H3 className={cx(titleStyle)}>{update.title}</H3>
              <Body className={cx(bodyStyle)}>{update.multi_line}</Body>
              <div className={cx(chipsStyle)}>
                {/* Display all tags from the three tag fields */}
                {[
                  ...(update.tags_category || []),
                  ...(update.tags_offerings || []),
                  ...(update.tags_product || []),
                ].map((tag: string) => (
                  <Chip key={tag} label={tag} className={cx(chipStyle)} />
                ))}
              </div>
            </Card>
          );
        })}
      </div>
    </div>
  );
};

export default Featured;
