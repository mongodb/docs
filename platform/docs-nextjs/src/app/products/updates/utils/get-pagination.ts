/**
 *
 * @param array - flatten array
 * @param chunkSize - max size of each array
 * @returns [[{}, {}], [{}]]
 * @description Takes an flatten array and creates one level nested array of objects based on the chunkSize. Used for pagination data.
 */
export const getPagination = <T>(array: T[], chunkSize: number): T[][] => {
  const chunks: T[][] = [];
  if (!array.length) {
    return [[]]; // default to one empty nested array
  }
  for (let i = 0; i < array.length; i += chunkSize) {
    chunks.push(array.slice(i, i + chunkSize));
  }
  return chunks;
};
