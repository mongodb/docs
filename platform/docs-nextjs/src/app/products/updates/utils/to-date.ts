export const normalizeDate = (date: string) => {
  // Check if the date string has the correct format 'YYYY-MM-DD'
  const dateToNormalize = /^(\d{4})-(\d{2})-(\d{2})$/.exec(date)?.slice(1); // excludes the full date in the first index i.e. "2025-05-27"
  if (dateToNormalize) {
    // This is needed because, JavaScript’s Date constructor uses zero-based months (January = 0 … December = 11).
    return new Date(Number(dateToNormalize[0]), Number(dateToNormalize[1]) - 1, Number(dateToNormalize[2]));
  }
  return new Date(date);
};
