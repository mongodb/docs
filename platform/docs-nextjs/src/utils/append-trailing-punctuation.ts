import { ASTNode, ReferenceNode, RefRoleNode } from "@/types/ast";
import { isReferenceNode, isRefRoleNode, isTextNode } from "@/types/ast-utils";

// Append dangling punctuation to preceding Link's text
export const appendTrailingPunctuation = (nodes: ASTNode[]) => {
  const truncatedNodes = [];

  for (let i = 0; i < nodes.length; i++) {
    const currNode = nodes[i];
    const nextNode = nodes[i + 1];
    const hasDanglingSibling =
      isTextNode(nextNode) &&
      nextNode.value.length === 1 &&
      !nextNode.value.match(/\s/);
    // TODO: if issues arise, instead of matching for non-whitespace, we might consider checking FOR certain singular punctuation
    // ie: regex match for  .?!,:;)*"']}`

    if ((isReferenceNode(currNode) || isRefRoleNode(currNode)) && hasDanglingSibling) {
      const copyOfNodeWithDeepCopiedChildren = deepCopyNodesChildren(currNode);
      copyOfNodeWithDeepCopiedChildren.children.push(nextNode);
      truncatedNodes.push(copyOfNodeWithDeepCopiedChildren);
      i++;
      continue;
    }
    truncatedNodes.push(currNode);
  }
  return truncatedNodes;
};

// Make a copy of node with a deep copy of the new child node
function deepCopyNodesChildren(refNode: ReferenceNode | RefRoleNode) {
  if (!refNode.children || !refNode.children.length) return { ...refNode };
  const copyOfChildren = [...refNode.children];

  const copyOfNode = {
    ...refNode,
    children: copyOfChildren,
  };
  return copyOfNode;
}
