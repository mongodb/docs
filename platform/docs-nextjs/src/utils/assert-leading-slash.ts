const assertLeadingSlash = (url: string) => {
  if (url && url.match(/^\//)) {
    return url;
  }
  return `/${url}`;
};

export { assertLeadingSlash };
