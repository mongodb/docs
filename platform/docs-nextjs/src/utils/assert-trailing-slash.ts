export const assertTrailingSlash = (url: string) => {
  if (url && url.match(/\/$/)) {
    return url;
  }
  return `${url}/`;
};
