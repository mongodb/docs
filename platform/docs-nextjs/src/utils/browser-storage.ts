import { isBrowser } from './is-browser';

const isValidStorage = isBrowser;

export const setLocalValue = (key: string, value: unknown) => {
  try {
    if (isValidStorage) {
      const prevState = JSON.parse(window.localStorage.getItem('mongodb-docs') || '{}');
      localStorage.setItem('mongodb-docs', JSON.stringify({ ...prevState, [key]: value }));
    }
  } catch {
    console.error('Error setting localStorage value');
  }
};

export const getLocalValue = (key: string) => {
  try {
    if (isValidStorage && JSON.parse(window.localStorage.getItem('mongodb-docs') || '{}')) {
      const docsObj = JSON.parse(window.localStorage.getItem('mongodb-docs') || '{}');
      if (docsObj) {
        return docsObj[key];
      }
    }
    return undefined;
  } catch {
    console.error('Error getting localStorage value');
    return undefined;
  }
};

export const setSessionValue = (key: string, value: string) => {
  try {
    if (isValidStorage) {
      const prevState = JSON.parse(window.sessionStorage.getItem('mongodb-docs') || '{}');
      sessionStorage.setItem('mongodb-docs', JSON.stringify({ ...prevState, [key]: value }));
    }
  } catch {
    console.error('Error setting session value');
  }
};

export const getSessionValue = (key: string) => {
  try {
    if (isValidStorage && JSON.parse(window.sessionStorage.getItem('mongodb-docs') || '{}')) {
      const docsObj = JSON.parse(window.sessionStorage.getItem('mongodb-docs') || '{}');
      if (docsObj) {
        return docsObj[key];
      }
    }
    return undefined;
  } catch {
    console.error('Error getting session value');
    return undefined;
  }
};
