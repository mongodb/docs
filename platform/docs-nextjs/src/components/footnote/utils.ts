import type { ASTNode } from '@/types/ast';
import type { FootnoteNode, FootnoteReferenceNode } from '@/types/ast';
import { findAllKeyValuePairs } from '@/utils/find-all-key-value-pairs';
import type { FootnoteType } from './footnote-context';

export const getFootnotes = (nodes: ASTNode[]) => {
  const footnotes = findAllKeyValuePairs(nodes, 'type', 'footnote') as FootnoteNode[];
  const footnoteReferences = findAllKeyValuePairs(nodes, 'type', 'footnote_reference') as FootnoteReferenceNode[];
  const numAnonRefs = footnoteReferences.filter(
    (node) => !Object.prototype.hasOwnProperty.call(node, 'refname'),
  ).length;
  // We label our footnotes by their index, regardless of their names to
  // circumvent cases such as [[1], [#], [2], ...]
  return footnotes.reduce<Record<string, FootnoteType>>((map, footnote, index) => {
    if (footnote.name) {
      // Find references associated with a named footnote

      map[footnote.name] = {
        label: index + 1,
        references: getNamedFootnoteReferences(footnoteReferences, footnote.name),
      };
    } else {
      // Find references associated with an anonymous footnote
      // Replace potentially broken anonymous footnote ids
      footnote.id = `${index + 1}`;

      map[footnote.id] = {
        label: index + 1,
        references: getAnonymousFootnoteReferences(index, numAnonRefs),
      };
    }
    return map;
  }, {});
};

// Find all footnote_reference node IDs associated with a given footnote by
// that footnote's refname
const getNamedFootnoteReferences = (footnoteReferences: FootnoteReferenceNode[], refname: string) => {
  return footnoteReferences.filter((node) => node.refname === refname).map((node) => node.id);
};

// They are used infrequently, but here we match an anonymous footnote to its reference.
// The nth footnote on a page is associated with the nth reference on the page. Since
// anon footnotes and footnote references are anonymous, we assume a 1:1 pairing, and
// have no need to query nodes. If there are more anonymous footnotes than references,
// we may return an empty array
const getAnonymousFootnoteReferences = (index: number, numAnonRefs: number) => {
  return index > numAnonRefs ? [] : [`id${index + 1}`];
};
