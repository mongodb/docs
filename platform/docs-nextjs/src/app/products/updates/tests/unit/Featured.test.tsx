import { render, screen } from '@testing-library/react';
import { useRouter } from 'next/navigation';
import Featured from '../../components/Featured';

import { mockedLimitedFeaturedUpdates } from '../utils/mock-featured-updates';

describe('Featured component', () => {
  beforeEach(() => {
    (useRouter as jest.Mock).mockReturnValue({ push: jest.fn() });
  });
  it('renders correctly', () => {
    const wrapper = render(<Featured updates={mockedLimitedFeaturedUpdates} />);
    const { getAllByTestId } = screen;
    const chips = getAllByTestId('chip-text');
    expect(wrapper.container.querySelector('h3')).toHaveTextContent('Featured Updates');
    expect(wrapper.queryAllByTestId('featured-update')).toHaveLength(3);
    expect(chips).toHaveLength(3);
    expect(wrapper.asFragment()).toMatchSnapshot();
  });
});
