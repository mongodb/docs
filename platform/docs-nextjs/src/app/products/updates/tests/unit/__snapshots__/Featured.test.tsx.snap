// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Featured component renders correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  max-width: 1440px;
  margin: 48px auto 0;
  padding: 0 24px;
  margin-bottom: 60px;
}

@media not all and (max-width: 767px) {
  .emotion-0 {
    padding: 0 32px;
  }
}

@media not all and (max-width: 1024px) {
  .emotion-0 {
    padding: 0 32px;
  }
}

@media not all and (max-width: 1440px) {
  .emotion-0 {
    padding: 0 8px;
  }
}

.emotion-1 {
  margin: unset;
  font-family: 'Euclid Circular A','Helvetica Neue',Helvetica,Arial,sans-serif;
  color: #001E2B;
  font-size: 24px;
  line-height: 32px;
  font-weight: 500;
  color: #001E2B;
  color: #000;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  gap: 16px;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  margin-top: 26px;
}

@media not all and (max-width: 767px) {
  .emotion-2 {
    gap: 16px;
  }
}

@media not all and (max-width: 1024px) {
  .emotion-2 {
    gap: 24px;
  }
}

.emotion-3 {
  position: relative;
  -webkit-transition: 150ms ease-in-out;
  transition: 150ms ease-in-out;
  transition-property: border,box-shadow;
  border-radius: 24px;
  font-family: 'Euclid Circular A','Helvetica Neue',Helvetica,Arial,sans-serif;
  font-size: 13px;
  line-height: 20px;
  padding: 24px;
  min-height: 68px;
  border: 1px solid #E8EDEB;
  box-shadow: 0 4px 10px -4px rgba(0,30,43,0.3);
  background-color: #FFFFFF;
  color: #1C2D38;
  cursor: pointer;
  -webkit-flex: 0 0 100%;
  -ms-flex: 0 0 100%;
  flex: 0 0 100%;
  min-width: 0;
  padding: 24px;
  min-height: 360px;
}

.emotion-3:focus {
  outline: none;
  box-shadow: 0 0 0 2px #FFFFFF,0 0 0 4px #0498EC,0 4px 10px -4px rgba(0,30,43,0.3);
}

.emotion-3:hover,
.emotion-3:active {
  border: 1px solid #E8EDEB;
  box-shadow: 0 4px 20px -4px rgba(0,30,43,0.2);
}

.emotion-3:hover:focus,
.emotion-3:active:focus {
  box-shadow: 0 0 0 2px #FFFFFF,0 0 0 4px #0498EC,0 4px 20px -4px rgba(0,30,43,0.2);
}

@media not all and (max-width: 767px) {
  .emotion-3 {
    -webkit-flex: 0 0 calc(50% - 8px);
    -ms-flex: 0 0 calc(50% - 8px);
    flex: 0 0 calc(50% - 8px);
    padding: 32px;
  }
}

@media not all and (max-width: 1024px) {
  .emotion-3 {
    -webkit-flex: 0 0 calc(33.333% - 16px);
    -ms-flex: 0 0 calc(33.333% - 16px);
    flex: 0 0 calc(33.333% - 16px);
    padding: 40px;
  }
}

.emotion-4 {
  margin: unset;
  font-family: 'Euclid Circular A','Helvetica Neue',Helvetica,Arial,sans-serif;
  color: #001E2B;
  font-size: 13px;
  line-height: 20px;
  color: #001E2B;
  font-weight: 400;
  font-size: 13px;
  line-height: 20px;
  color: var(--whats-new-gray-dark1);
  margin-bottom: 16px;
}

.emotion-4 strong,
.emotion-4 b {
  font-weight: 600;
}

.emotion-5 {
  margin: unset;
  font-family: 'Euclid Circular A','Helvetica Neue',Helvetica,Arial,sans-serif;
  color: #001E2B;
  font-size: 24px;
  line-height: 32px;
  font-weight: 500;
  color: #001E2B;
  font-size: 24px;
  line-height: 32px;
  margin-bottom: 8px;
  color: var(--whats-new-black);
}

.emotion-6 {
  margin: unset;
  font-family: 'Euclid Circular A','Helvetica Neue',Helvetica,Arial,sans-serif;
  color: #001E2B;
  font-size: 13px;
  line-height: 20px;
  color: #001E2B;
  font-weight: 400;
  font-size: 16px;
  line-height: 24px;
  color: var(--whats-new-black);
  margin-bottom: 44px;
  display: -webkit-box;
  -webkit-line-clamp: 3;
  -webkit-box-orient: vertical;
  overflow: hidden;
}

.emotion-6 strong,
.emotion-6 b {
  font-weight: 600;
}

.emotion-7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  gap: 8px;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.emotion-8 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  overflow: hidden;
  white-space: nowrap;
  border-radius: 4px;
  font-size: 13px;
  line-height: 18px;
  color: #0C2657;
  background-color: #E1F7FF;
  -webkit-transition: background-color 100ms ease-in-out;
  transition: background-color 100ms ease-in-out;
  font-size: 16px;
  padding: 2px;
}

.emotion-9 {
  padding-inline: 4px;
  padding-block: 0px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 2px;
}

.emotion-9 svg {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  color: #083C90;
}

.emotion-9:focus-within {
  background-color: #89D2FF;
}

.emotion-9 .lg-ui-chip-inline-definition-0000:focus-visible,
.emotion-9 .lg-ui-chip-inline-definition-0000:focus {
  outline: none;
}

<div
    class="emotion-0"
  >
    <h3
      class="emotion-1"
    >
      Featured Updates
    </h3>
    <div
      class="emotion-2"
    >
      <div
        class="emotion-3"
        data-testid="featured-update"
      >
        <p
          class="emotion-4"
        >
          September 17, 2025
        </p>
        <h3
          class="emotion-5"
        >
          Back again for the win CB
        </h3>
        <p
          class="emotion-6"
        >
          This is a test from CB
        </p>
        <div
          class="emotion-7"
        >
          <span
            aria-disabled="false"
            class="emotion-8"
          >
            <span
              class="emotion-9 lg-ui-chip-text-0000"
              data-testid="chip-text"
            >
              foo
            </span>
          </span>
          <span
            aria-disabled="false"
            class="emotion-8"
          >
            <span
              class="emotion-9 lg-ui-chip-text-0000"
              data-testid="chip-text"
            >
              bar
            </span>
          </span>
          <span
            aria-disabled="false"
            class="emotion-8"
          >
            <span
              class="emotion-9 lg-ui-chip-text-0000"
              data-testid="chip-text"
            >
              baz
            </span>
          </span>
        </div>
      </div>
      <div
        class="emotion-3"
        data-testid="featured-update"
      >
        <p
          class="emotion-4"
        >
          September 17, 2025
        </p>
        <h3
          class="emotion-5"
        >
          Final CB test dev entry
        </h3>
        <p
          class="emotion-6"
        >
          This is a test from CB
        </p>
        <div
          class="emotion-7"
        />
      </div>
      <div
        class="emotion-3"
        data-testid="featured-update"
      >
        <p
          class="emotion-4"
        >
          September 17, 2025
        </p>
        <h3
          class="emotion-5"
        >
          Another CB test dev entry
        </h3>
        <p
          class="emotion-6"
        >
          This is a test from CB
        </p>
        <div
          class="emotion-7"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
